src/runtime/dsl.ts

```diff
@@ -0,0 +1,431 @@
+//--------------------------------------------------------------------
+// Javascript DSL for writing Eve programs
+//--------------------------------------------------------------------
+
+// There don't seem to be TypeScript definitions for these by default,
+// so here we are.
+declare var Proxy:new (obj:any, proxy:any) => any;
+declare var Symbol:any;
+
+import {RawValue, Register, isRegister, GlobalInterner, Scan, IGNORE_REG, ID,
+        InsertNode, Node, Constraint, FunctionConstraint} from "./runtime";
+import * as runtime from "./runtime";
+import * as indexes from "./indexes";
+
+//--------------------------------------------------------------------
+// Utils
+//--------------------------------------------------------------------
+
+function maybeIntern(value:(RawValue|Register)):Register|ID {
+  if(value === undefined || value === null) throw new Error("Trying to intern an undefined");
+  if(isRegister(value)) return value;
+  return GlobalInterner.intern(value);
+}
+
+//--------------------------------------------------------------------
+// DSLVariable
+//--------------------------------------------------------------------
+
+type DSLValue = RawValue|Register;
+class DSLVariable {
+  static CURRENT_ID = 0;
+  id: number;
+  constructor(public name:string, public value:DSLValue = new Register(-1)) {
```

Comment:
Seems like it's used intentionally that way elsewhere, which means it's not an error case but a weird sentinel value? We should consider expanding the type definition rather than having a malformed register.